### PR TITLE
Handle `isdefined` exprs

### PR DESCRIPTION
**Pull request by TH3CHARLie**

### CODE REVIEW

**Comment by TH3CHARLie** [comment]
I am worried that the testcase may not be representative enough though. The testcase will pass before the changes because we have the rule that `isbitstype` will not propagate escape information and `isdefined` returns a boolean which matches this rule.

File: test/runtests.jl [hunk 1]
```diff
@@ -283,6 +283,14 @@ end
         @test !isnothing(i)
         @test is_no_escape(escapes.ssavalues[i])
     end
+
+    let
+        src, escapes = analyze_escapes((String,)) do s
+            b = isdefined(s, :value)
+            global bb = b
+        end
```
Comment (aviatesk) [inline test/runtests.jl]: `isdefined` here is just a builtin function call.
`:isdefind` expression will be introduced by `@isdefined` macro or other program constructs:
```julia
analyze_escapes((Ref{String},)) do a
    return @isdefined(b)
end
```

**Review by aviatesk — CHANGES_REQUESTED** [review]
What is the original motivating example ? How `:isdefined` expression is used within the example ?

File: test/runtests.jl (hunk 1 above)
Comment (TH3CHARLie) [reply test/runtests.jl]: Oh that's my miss. update it now

File: test/runtests.jl [hunk 2]
```diff
@@ -283,6 +283,13 @@ end
         @test !isnothing(i)
         @test is_no_escape(escapes.ssavalues[i])
     end
+
+    let
+        src, escapes = analyze_escapes((Ref{String},)) do a
+            return @isdefined(b)
```
Comment (aviatesk) [inline test/runtests.jl]: Hm, I should have said in this case the optimizer should have elided this `:isdefined` check already at this point..
I'm wondering what are the cases we really need to handle.

Comment (TH3CHARLie) [reply test/runtests.jl]: let me record the exprs you presented in #23 and see what's the case of `isdefined`.

Comment (aviatesk) [reply test/runtests.jl]: Yeah, that would be a good starting point of investigation.

Comment (TH3CHARLie) [reply test/runtests.jl]: I did the experiment and also verify the testcase, finding that actually in this case the `:isdefined` expr is not elided:
```
julia> using EscapeAnalysis

julia> function f(a::String)
           return @isdefined(b)
       end
f (generic function with 1 method)

julia> src, escapes = @analyze_escapes f("haha")
typeof(f)(_2::String ↑)
1 ◌ 1 ─ %1 = $(Expr(:isdefined, :(Main.b)))::Bool
2 ◌ └──      return %1
```

Comment (aviatesk) [reply test/runtests.jl]: If `:isdefined` only deals with global objects, then it's not our interest ? When `:isdefined` works on objects within local scope ?

Comment (TH3CHARLie) [reply test/runtests.jl]: There are two types of `isdefined` in my collected results of #23 :
```
- [ ] `:$(Expr(:isdefined, :(%5)))`
- [ ] `:$(Expr(:isdefined, :($(Expr(:static_parameter, 1)))))`
```
the first one is clearly an `isdefined` on local scope, the second I am not sure about it.

Comment (aviatesk) [inline test/runtests.jl]:
```suggestion
            return @isdefined(a) # XXX `:isdefined` expression should have been elided in this case
```

Comment (aviatesk) [reply test/runtests.jl]: We may want to have either of those cases as a test case here, but it's also obvious that `:isdefined` never escapes anything, so I'm okay to merge this as is if you want.

**Review by aviatesk — APPROVED** [review]

File: test/runtests.jl (hunk 2 above)
Comment (TH3CHARLie) [reply test/runtests.jl]: I will add another case to cover the local scope case.

Comment (TH3CHARLie) [reply test/runtests.jl]: added

**Review by aviatesk — APPROVED** [review]
Great work !

**Pull request merged**